File: 2053/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public String kthDistinct(String[] arr, int k) {
        Map<String, Integer> counts = new HashMap<>();
        for (String s : arr) {
            counts.put(s, counts.getOrDefault(s, 0) + 1);
        }
        for (String s : arr) {
            if (counts.get(s) == 1) {
                k--;
            }
            if (k == 0) {
                return s;
            }
        }
        return "";
    }
}

// time O(n)
// space O(n)

File: 2053/solution.ts
function kthDistinct(arr: string[], k: number): string {
  const counts = new Map<String, number>();
  for (const s of arr) {
    counts.set(s, (counts.get(s) ?? 0) + 1);
  }

  for (const s of arr) {
    if (counts.get(s) === 1) {
      k--;
    }
    if (k === 0) {
      return s;
    }
  }
  return "";
}
